File: CSC455-Game-Programming/Pathfinder/core/src/main/kotlin/com/jeff/game/pathfinder/graph/Node.kt
package com.jeff.game.pathfinder.graph

import com.jeff.game.pathfinder.util.option.None
import com.jeff.game.pathfinder.util.option.Option
import com.jeff.game.pathfinder.util.option.Some
import java.util.*

/**
 * Class to represent a node.
 * @constructor Default constructor
 */
class Node(val cost: Int) {

    var marker: Option<Marker> = None()
    var edges: Option<List<Edge>> = None()
        set(value) {
            if (!field.defined()) {
                field = value
            } else {
                throw IllegalStateException()
            }
        }

    /**
     * Method to return list of cells connected from this node by an edge.
     */
    fun connectedCells(): Option<List<Node>> {
        if (edges.defined()) {
            val builder = ArrayList<Node>()
            edges.get().forEach { builder.add(it.to) }
            return Some(builder.toList())
        } else {
            return None()
        }
    }

}

File: CSC455-Game-Programming/Chaser/core/src/com/jeff/chaser/models/graph/Node.scala
package com.jeff.chaser.models.graph


class Node(val cost: Float, val cX: Float, val cY: Float, val fromX: Float, val toX: Float,
           val fromY: Float, val toY: Float) {

  var edges: List[Edge] = _
  var marker: Marker = _

  def within(x: Float, y: Float): Boolean = (x >= fromX && x <= toX) && (y >= fromY && y <= toY)

}
